Detailed error responses when missing fields.

vqtran opened the pull request:
Refactored the QueryResult object to be truer to its name.
Querier now does a validation step that walks over a QueryRequest and
checks if all expected fields are there. If not, it makes Alexa announce
a useful error message.

Next step re handling errors:
  - Make these error responses interactive. Alexa will try to
    interactively fix invalid requests with you.
  - Fix up SchemaInferrer / errors where column/table doesn't exist.
    For example, right now:

    how many sales where product is speakers and counts are nine

    fails because SchemaInferrer tries to look up counts but it doesn't
    exist, only count does. This null pointers, when it really should
    propagate some error upwards back to the user.

vqtran commented:
Also this isn't meant to be interactive yet, these changes felt big enough for one PR

GabeLoins commented on src/main/java/echoquery/intents/AggregationHandler.java:
```diff
@@ -34,12 +34,6 @@ public SpeechletResponse respond(Intent intent, Session session) {
-    try {
-      QueryResult result = querier.execute(QueryRequest.of(intent));
-      return result.toEnglish();
-    } catch (SQLException e) {
```
I will miss runtime exceptions being an integral component of our basic error handling

vqtran replied:
Now it'll just be a sneaky component!

GabeLoins commented on src/main/java/echoquery/sql/Querier.java:
```diff
@@ -21,11 +24,98 @@ public Querier(Connection conn) {
+      return Optional.of(new QueryResult(Status.FAILURE,
+          "I can't tell if you want to know the number of rows or the average, "
+          + "sum, min, or max of a particular column in the "
+          + request.getFromTable().get() + " table. Please try again."));
```
should we just default to count in this case? just a thought

vqtran replied:
this sounded weird at first but i think it makes sense, yeah

GabeLoins commented on src/main/java/echoquery/sql/Querier.java:
```diff
@@ -21,11 +24,98 @@ public Querier(Connection conn) {
+            + "or or connecting them, and I'm not sure what to do with this."
+            + " Please try again."));
+      }
+    }
```
MEGAMETHOD

vqtran replied:
i'll helper method this, although helper methods without reuse purpose is kind of meh

GabeLoins replied:
please dont I think its pretty readable right now I was just messing around

GabeLoins commented on src/main/java/echoquery/sql/Querier.java:
```diff
@@ -21,11 +24,98 @@ public Querier(Connection conn) {
-  public QueryResult execute(QueryRequest request) throws SQLException {
+  public QueryResult execute(QueryRequest request) {
+    // First validate the request, interactively mending malformed requests.
+    Optional<QueryResult> invalidation = validate(request);
```
should validate be in query request?

vqtran replied:
I think having a validator validating requests makes more sense than every request validating itself, more of a pipeline

GabeLoins commented:
looks good !

GabeLoins commented:
you can merge

vqtran commented:
Changed default to count, added comments for QueryRequest, Querier, QueryResult since I think they're pretty stable now, refactored tests to cut out like 200 lines
